`MenuItemsChoice`: Refactor to TypeScript

flootr opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Refactor `MenuItemsChoice` component to TypeScript

## Why?

Part of https://github.com/WordPress/gutenberg/issues/35744

## How?

- Followed general recommendations from the [TypeScript Migration guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#refactoring-a-component-to-typescript). 
- Renamed file to `.tsx` and created a `types.ts` file with relevant types

## Testing Instructions

1. Ensure there are no type errors
2. Component should behave as expected

Wit this PR we introduce a story for this component. Please verify it behaves as expected.

flootr committed:
`MenuItemsChoice`: Refactor to TypeScript

flootr committed:
create stories for `MenuItemsChoice` /wip

flootr committed:
enhance story with interactivity

mirka commented on packages/components/src/menu-items-choice/types.ts:
```diff
@@ -0,0 +1,50 @@
+
+export type MenuItemsChoiceProps = {
+	/**
+	 * Array of choices
```
Just as a style consistency thing, we end all descriptions with a period.

```suggestion
	 * Array of choices.
```

mirka commented on packages/components/src/menu-items-choice/types.ts:
```diff
@@ -0,0 +1,50 @@
+	/**
+	 * Array of choices
+	 */
+	choices: readonly MenuItemChoice[];
```
Interesting! I've never really considered using `readonly` like this because it's a given that a React component won't mutate its props, but I guess it adds another layer of protection. Just curious, have you ever encountered situations where a component was inadvertently mutating a prop?

cc @ciampo

ciampo replied:
Never personally seen anything like it, off the top of my mind. But curiously enough, I encountered recently `readonly` in the `QueryControls` typescript refactor PR: https://github.com/WordPress/gutenberg/pull/46721/#discussion_r1055199840

flootr replied:
To be honest this is a [copy paste](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wordpress__components/menu-items-choice/index.d.ts#L10). I'll let you decide whether that's something we want to keep.

I've definitely seen code which reassigns and/or mutates props but it's not exactly considered a good practice and is usually avoided. As you mentioned I'd consider `readonly` an additional layer of protection which warns you when you're trying to e.g. `.push` something to the `choices` array instead of creating a new array and spread the existing array.

mirka replied:
I mean, it certainly doesn't hurt! I looked into it and some people actually suggest doing `readonly` by default on all  arrays 😆 I will keep it in mind as well.

mirka commented on packages/components/src/menu-items-choice/stories/index.tsx:
```diff
@@ -0,0 +1,70 @@
+export default meta;
+
+const Template: ComponentStory< typeof MenuItemsChoice > = ( args ) => {
+	const [ choice, setChoice ] = useState( args.value ?? '' );
```
For controlled components we usually just disable the control for `value` to avoid making things murky:

https://github.com/WordPress/gutenberg/blob/5115017b4129e942cce903a334f9d0350e21a0fe/packages/components/src/select-control/stories/index.tsx#L25

And then we can do 

```suggestion
	const [ choice, setChoice ] = useState( args.choices[ 0 ]?.value );
```

or something like that.

mirka commented on packages/components/src/menu-items-choice/stories/index.tsx:
```diff
@@ -0,0 +1,70 @@
+				onSelect={ ( value ) => setChoice( value ) }
+				onHover={ ( value ) => setHovered( value ) }
+			/>
+			<div>Currently hovered value: { hovered ? hovered : '-' }</div>
```
For callbacks like the `onHover` here, we usually use the Storybook [Actions](https://storybook.js.org/docs/react/essentials/actions) feature so we don't have to render callback indicators like "Currently hovered...". Here's an example implementation:

https://github.com/WordPress/gutenberg/blob/5115017b4129e942cce903a334f9d0350e21a0fe/packages/components/src/search-control/stories/index.tsx#L19-L21

https://github.com/WordPress/gutenberg/blob/5115017b4129e942cce903a334f9d0350e21a0fe/packages/components/src/search-control/stories/index.tsx#L39-L42

flootr replied:
TIL! Thanks. Changed it via https://github.com/WordPress/gutenberg/pull/47180/commits/b46f7706f371b30742b4ccb93fac7dae8b69a605.

mirka commented on packages/components/src/menu-items-choice/index.tsx:
```diff
@@ -0,0 +1,50 @@
+	);
+}
+
+export default MenuItemsChoice;
```
We'll also want to add a JSDoc on the main export with a short description and a code snippet like [this](https://github.com/WordPress/gutenberg/blob/43d9c82922619c1d1ff6b454f86f75c3157d3de6/packages/components/src/date-time/date-time/index.tsx#L193-L217). We usually copy paste what's already in the readme file.

mirka commented on packages/components/src/menu-items-choice/index.tsx:
```diff
@@ -0,0 +1,50 @@
+							}
+						} }
+						onMouseEnter={ () => onHover( item.value ) }
+						onMouseLeave={ () => onHover() }
```
Any reason why this was changed from `() => onHover( null )`? This could be a breaking change for a consumer that was specifically doing `null` checks, so I think we should avoid the change if possible.

flootr replied:
I understand we need to keep the API to make sure we don't introduce a breaking change at this point. There's no particular reason other than me finding this kind of an odd API to return explicitly `null` in case no element is hovered. Reverted via https://github.com/WordPress/gutenberg/pull/47180/commits/9baa7fb26a56a3c386f8fa70f49ad3fb95d8eb4e.

mirka reviewed (commented):
Thank you for helping out with the TS migrations! I appreciate you adding a Storybook story so we can see what's going on.

We'll also need a brief [changelog](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) entry, thanks 🙏

flootr committed:
improve story for `MenuItemsChoice`

flootr committed:
end all type descriptions with a dot

flootr committed:
`MenuItemsChoice`: add description to default export

flootr committed:
`MenuItemsChoice.onHover`: revert to returning `null` onmouseleave

flootr committed:
add changelog entry

mirka commented on packages/components/src/menu-items-choice/index.tsx:
```diff
@@ -0,0 +1,83 @@
+	);
+}
+
+/**
+ * `MenuItemsChoice` functions similarly to a set of `MenuItem`s, but allows the user to select one option from a set of multiple choices.
+ *
```
Looks like this JSDoc needs to be placed above the function declaration (L13), not here above the export statement.

When the JSDoc is associated correctly, it will show up in the Storybook Docs view, and also in IntelliSense when you hover over the component in an IDE.

flootr replied:
✅ https://github.com/WordPress/gutenberg/pull/47180/commits/c7273984ac674b7c157f130f0c0b4c6ce783e6b8

mirka reviewed (approved):
Thanks! The JSDoc placement needs to be tweaked, but otherwise this looks ready to merge 🚀

flootr committed:
move jsdoc example above component decl

This is necessary so it's shown on the documentation page of the Storybook entry.

ciampo commented on packages/components/src/menu-items-choice/types.ts:
```diff
@@ -0,0 +1,50 @@
+import type { ShortcutProps } from '../shortcut/types';
+
+export type MenuItemsChoiceProps = {
+	/**
+	 * Array of choices.
+	 */
```
```suggestion
	/**
	 * Array of choices.
	 *
	 * @default []
	 */
```

ciampo commented on packages/components/src/menu-items-choice/types.ts:
```diff
@@ -0,0 +1,50 @@
+	 * selects a new choice.
+	 */
+	onSelect( value: string ): void;
+	/**
+	 * Callback function to be called with a choice when user
+	 * hovers over a new choice (will be empty on mouse leave).
+	 */
+	onHover: ( value: string | null ) => void;
```
```suggestion
	/**
	 * Callback function to be called with a choice when user
	 * hovers over a new choice (will be empty on mouse leave).
	 *
	 * @default noop
	 */
	onHover: ( value: string | null ) => void;
```

ciampo reviewed (commented):
Left a couple of minor comments about missing `@default` tags in the JSDoc descriptions.

I also noticed that this component doesn't list any of its props in the README, which means that we don't really have to amend that as part of this PR, if we want to speed up the TS refactor

flootr committed:
types: add missing `@default` tags

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

ciampo committed:
Move CHANGELOG entry to unreleased section
